crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -898,16 +900,25 @@ impl RecoveryContextKind {
                         | TokenKind::Lpar
                 )
             }
-            // The names of an import statement cannot be parenthesized, so it
-            // always ends with a newline.
-            RecoveryContextKind::ImportNames => p.at(TokenKind::Newline),
+            // The names of an import statement cannot be parenthesized, so `)` is not a
+            // terminator.
+            RecoveryContextKind::ImportNames => {
+                // test_ok import_stmt_terminator
+                // import a, b; import c, d
+                // import a, b
+                // c, d
+                matches!(p.current_token_kind(), TokenKind::Semi | TokenKind::Newline)
```

Comment:
Adding the `Semi` only to some of the recovery context feels a bit strange to me. What about if there's a semicolon in an array expression, shouldn't we not parse the semicolon there and consider what comes after to be a simple statement? 

It does feel kind of wrong but I wonder if `is_list_element` should return `true` when parsing a body. It feels wrong because `;` isn't actually a list element but it does mark the start of the next statement.

Reply:
I'm not a big fan of this either but the reason it exists only for a subset of recovery context is that the list of nodes being parsed is part of a simple statement _and_ is at the end of the statement. For example, `import <names>`.

I tried thinking of using list parsing for sequence of simple statements separated by semicolons (`parse_simple_statements`) but that also doesn't work. Should we use `;` as the terminator token? But that would still mean that the inner context got an unexpected token instead of a comma.